homework2/hw1.py

```diff
@@ -0,0 +1,122 @@
+"""
+Given a file containing text. Complete using only default collections:
+    1) Find 10 longest words consisting from largest amount of unique symbols
+    2) Find rarest symbol for document
+    3) Count every punctuation char
+    4) Count every non ascii char
+    5) Find most common non ascii char for document
+"""
+from typing import List
+from math import inf
+
+
+def get_longest_diverse_words(file_path: str) -> List[str]:
+    """
+    Find 10 longest words consisting from largest amount of unique symbols
+    :param file_path: name and path of text file
+    :return: list of the longest words
+    """
+    dict_length = {}
+    with open(file_path, encoding='unicode_escape', errors="ignore") as fi:
+        for line in fi:
+            line_list = line.split()
+            for word in line_list:
+                if dict_length.get(len(set(word))) is None:
+                    dict_length[len(set(word))] = [word]
+                else:
+                    dict_length[len(set(word))].append(word)
+    list_key_sorted = sorted(dict_length, reverse=True)
+    list_resulted = []
+    for key in list_key_sorted:
+        while dict_length[key] and len(list_resulted) < 10:
+            list_resulted.append(dict_length[key].pop(0))
+        else:
+            continue
+    return list_resulted
+
+
+def get_rarest_char(file_path: str) -> str:
+    """
+    Find the rarest symbol for document, take first symbol if the same value
+    :param file_path: name and path of text file
+    :return: the rarest symbols in text
+    """
+    dict_symbols = {}
+    with open(file_path, encoding='unicode_escape', errors="ignore") as fi:
+        for line in fi:
+            line_list = line.split()
+            for word in line_list:
+                for symbol in word:
```

Comment:
this function has O(N^3) complexity - it's too complex. try to simplify it. we don't actually need 3 nested for-loops here. try not to split text by words